data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/S3Source.java

```diff
@@ -47,7 +50,35 @@ public void start(Buffer<Record<Event>> buffer) {
         final ConfigBucketOwnerProviderFactory configBucketOwnerProviderFactory = new ConfigBucketOwnerProviderFactory();
         final BucketOwnerProvider bucketOwnerProvider = configBucketOwnerProviderFactory.createBucketOwnerProvider(s3SourceConfig);
 
-        final S3Service s3Service = new S3Service(s3SourceConfig, buffer, codec, pluginMetrics, bucketOwnerProvider);
+        Optional<S3SelectOptions> s3SelectOptional = Optional.ofNullable(s3SourceConfig.getS3SelectOptions());
+        S3ObjectPluginMetrics s3ObjectPluginMetrics = new S3ObjectPluginMetrics(pluginMetrics);
+
+        S3ClientBuilderFactory s3ClientBuilderFactory = new S3ClientBuilderFactory(s3SourceConfig);
+        final S3ObjectHandler s3Handler;
+        final S3ObjectRequest.Builder s3ObjectRequestBuilder = new S3ObjectRequest.Builder(buffer, s3SourceConfig.getNumberOfRecordsToAccumulate(),
+                s3SourceConfig.getBufferTimeout(), s3ObjectPluginMetrics);
+        final BiConsumer<Event, S3ObjectReference> eventMetadataModifier = new EventMetadataModifier(
+                s3SourceConfig.getMetadataRootKey());
+        if (s3SelectOptional.isPresent()) {
+            S3ObjectRequest s3ObjectRequest = s3ObjectRequestBuilder.queryStatement(s3SelectOptional.get().getQueryStatement())
+                    .serializationFormatOption(s3SelectOptional.get().getS3SelectSerializationFormatOption())
+                    .s3AsyncClient(s3ClientBuilderFactory.getS3AsyncClient()).eventConsumer(eventMetadataModifier).
+                    fileHeaderInfo(FileHeaderInfo.valueOf(s3SelectOptional.get().getCsvFileHeaderInfo().toUpperCase()))
+                    .compressionType(CompressionType.valueOf(s3SourceConfig.getCompression().name()))
+                    .s3SelectResponseHandler(new S3SelectResponseHandler()).build();
+            s3Handler = new S3SelectObjectWorker(s3ObjectRequest);
+        } else {
+            if(s3SourceConfig.getCodec() == null)
+                throw new NoSuchElementException("codec is required in pipeline yaml configuration");
```

Comment:
This is not a good exception for this error.

You actually should be able to remove this condition.

Update the `S3SourceConfig` class with an `@AssertTrue` statement.

It will probably look like:

```
@AssertTrue(message = "A codec is required for reading objects.)
boolean isCodecProvidedWhenNeeded() {
  if(s3SelectOptions == null)
     return codec != null;
  return true;
}
```

Here is an example from another project:

https://github.com/opensearch-project/data-prepper/blob/298e7931aa3b26130048ac3bde260e066857df54/data-prepper-plugins/otel-trace-source/src/main/java/org/opensearch/dataprepper/plugins/source/oteltrace/OTelTraceSourceConfig.java#L101C5-L104

Reply:
Thanks David for your review, Changes are incorporated and available for review.